toe/src/urbit/app/toe.hoon

```diff
@@ -0,0 +1,921 @@
+    ^-  (list (list spot))
+    :~  ~[[%1 %1] [%2 %1] [%3 %1]]
+        ~[[%1 %2] [%2 %2] [%3 %2]]
+        ~[[%1 %3] [%2 %3] [%3 %3]]
+        ~[[%1 %1] [%1 %2] [%1 %3]]
+        ~[[%2 %1] [%2 %2] [%2 %3]]
+        ~[[%3 %1] [%3 %2] [%3 %3]]
+        ~[[%1 %3] [%2 %2] [%3 %1]]
+        ~[[%1 %1] [%2 %2] [%3 %3]]
+    ==
+  --
+::  +connect: handles subscriptions and updates
+::
+++  connect
+  |=  pax=path
+  ^-  (quip card _state)
+  ~&  who-connect+pax
+  |^  ?+    pax  ~|([%peer-toe-strange pax] !!)
+          [%sole *]           sole
+          [%toetile ~]        tile
+          [%room ^]           (room i.t.pax)
+          [%http-response *]  [~ state]
+      ==
+  ::  +tile: connects to the frontend
+  ::
+  ++  tile
+    ^-  (quip card _state)
+    ~&  tiled+src.bol
+    :_  state
+    [%give %fact ~ %json !>(*json)]~
+  ::  +sole: innitializes the console's state
+  ::
+  ++  sole
+    ^-  (quip card _state)
+    ~&  sole+[me src.bol]
+    ?.  (team:title me src.bol)  ~|([%strange-sole src.bol] !!)
+    :_  state(state.cli *sole-share)
+    =,  co:view
+    [(send mor+~[clear welcome wopr grid shall-we (prompt choose)])]~
+  ::  +room: receives a subscription to connecto to a room
+  ::
+  ++  room
+    |=  id=@t
+    ^-  (quip card _state)
+    =,  room-core
+    ?~  rooms  [~ state]
+    =/  ze=@p  (slav %p id)
+    ~&  [ze ship:current]
+    ?.  =(ze ship:current)  [~ state]
+    ?:  =(%wait game)
+      (join ze)
+    ?.  =(%play game)  [~ state]
+    (start ze)
+  --
+::  +view
+::  TODO: please make me into a %view app!
+::
+++  view
+  |%
+  ::  TODO: move other view updates here
+  ::
+  ++  updates
+    |%
+    ++  restart
+      ^-  (list card)
+      :~  (send:co mor+all-effects:co)
+          (send:fe [%status s+'null']~)
+      ==
+    ::
+    ++  receive
+      |=  guest=@p
+      =/  room-name=tape  (cite:title guest)
+      ?^  rooms
+        (notify guest)
+      :~  (send:fe ~[[%message s+(crip room-name)] [%status s+'confirm']])
+          =,  co
+          (send (prompt "{confirm}{room-name}? (Y/N) | "))
+      ==
+    ::
+    ++  confirm
+      =/  next=@p  guest:next:room-core
+      :~  (confirm:updates:fe next)
+          (confirm:updates:co next)
+      ==
+    ::
+    ++  playing
+      |=  move=toe-turno
+      ^-  (list card)
+      =/  [ze=@p who=@p]   =,  current:room-core
+        [ship who]
+      :~  ::  If the move originated in the frontend, this is redundant
+          ::  so this move is ignored on the frontend (via %next)
+          ::
+          %-  send:fe  ^-  (list [@t json])
+          :~  [%status s+'play']
+              [%next s+(scot %p who)]
+              [%stone s+`@t`stone.per.move]
+              [%move a+~[n+(scot %u -.spo.move) n+(scot %u +.spo.move)]]
+          ==
+        ::
+          =,  co
+          %-  send  ^-  sole-effect
+          :~  %mor
+              grid
+              %-  prompt
+              %-  dial
+              [ze stones:room-core ?:(=(who me) "<-" "->")]
```

Comment:
Similar to my other question about passing tapes: why not just pass the ship name, or a flag of some sort?